wgengine/magicsock/endpoint.go

```diff
@@ -354,6 +359,44 @@ func (de *endpoint) addrForWireGuardSendLocked(now mono.Time) (udpAddr netip.Add
 	return udpAddr, needPing
 }
 
+// addrForPingSizeLocked returns the address(es) that should be used for sending
+// the next ping. It will only return addrs with a large enough path MTU to
+// permit a ping payload of size bytes to be delivered (DERP is always one such
+// addr as it is a TCP connection). If it returns a zero-value udpAddr, then we
+// should continue probing the MTU of all paths to this endpoint. Zero, one, or
+// both of the returned UDP address and DERP address may be non-zero.
+//
+// de.mu must be held.
+func (de *endpoint) addrForPingSizeLocked(now mono.Time, size int) (udpAddr, derpAddr netip.AddrPort) {
+	if size == 0 {
+		udpAddr, derpAddr, _ = de.addrForSendLocked(now)
+		return
+	}
+
+	udpAddr = de.bestAddr.AddrPort
+	mtuOk := true
+	pathMTU := de.bestAddr.wireMTU
+	requestedMTU := pingSizeToPktLen(size, udpAddr.Addr())
+	if requestedMTU > pathMTU {
+		mtuOk = false
+	}
+
+	if udpAddr.IsValid() && mtuOk {
+		if !now.After(de.trustBestAddrUntil) {
```

Comment:
There's another check for `de.trustBestAddrUntil` in `cliPing` (the only caller of this new function), so perhaps we could simplify and only check it once?

Reply:
It is possible but I've spent 20 minutes trying to explain and the short version is: I want to keep the logic flow as similar to the other 2 versions of this function so that when we rewrite them all to share code in a way that makes some kind of sense, it's easier to do. For this PR, I want to do the minimum amount of risky change, and this is the version that is most like the original.